engine/gui/widgets.py

```diff
@@ -106,10 +108,86 @@ def create(
                 None,
             )
 
-        return GUISpec(
+        gui_spec = GUISpec(
             assets=[Asset(**a) for a in spec.get("assets", [])],
             buttons=buttons,
             images=[Image(**b) for b in spec.get("images", [])],
             cancel_action=cancel_action,
             initial_selected_button=selected_button,
         )
+
+        cls.validate(gui_spec)
+        return gui_spec
+
+    def validate(self) -> None:
+        """Validates the GUISpec."""
+
+        # Validation errors for images have not been tested
+        if not self.assets:
+            raise ValidationError("Must have at least one asset")
+        if not self.buttons and not self.images:
+            raise ValidationError("Must have at least one button or image")
+
+        duplicate_assets = [
+            asset for asset, count in Counter(self.assets).items() if count > 1
+        ]
+        if len(duplicate_assets) > 0:
+            raise ValidationError(f"Duplicate asset name(s): {duplicate_assets}")
+
+        if len(set(self.assets)) != len(set(self.buttons)) + len(set(self.images)):
```

Comment:
You typically don't want to use `set()` on objects unless you have a reasonable assurance that they will never change in the future (for example, a `Point2D` class is unlikely to ever change). The reason is that sets can only contain immutable types, and objects may or may not contain mutable types. Even if they don't contain mutable types right now, they may contain mutable types in the future, which means that someone modifying the class may inadvertently break this code (this is called "brittleness").

In this case it doesn't really matter since the intention is for `GUISpec` to be immutable, but I figured I'd raise this point.

More importantly though the logic is not entirely correct, it'd break if there are two buttons or images using the same asset. It's not hard to imagine someone wanting to re-use an asset for multiple images, and I suppose it's possible someone may want to re-use an asset for a button - there's no technical reason for us _not_ to allow that, so I'd say there's no real reason to enforce a constraint on it.

Since the intention of this block of code is to ensure that all the assets specified by buttons and images exist in `assets`, consider this alternative approach using sets (note the use of [set comprehensions](https://python-reference.readthedocs.io/en/latest/docs/comprehensions/set_comprehension.html)):
```python
assets_used = (
  {b.selected_image_asset for b in self.buttons} |
  {b.unselected_image_asset for b in self.buttons} |
  {img.image_asset for img in self.images}
)
assets_defined = {a.name for a in self.assets}

undefined_assets = assets_used - assets_defined
```
This would have the advantage of being less code and it's more efficient (it's linear instead of quadratic), not that efficiency really matters at the scale this would be used at.